### PR TITLE
Allow duplicate entries in the DB.

### PR DESCRIPTION
**Pull request by AlCutter**
Mainly to allow for mirroring other logs which do allow duplicate submissions.

### CODE REVIEW

File: cpp/log/sqlite_db-inl.h [hunk 1]
```diff
@@ -164,24 +168,7 @@ typename Database<Logged>::WriteResult SQLiteDB<Logged>::CreateSequencedEntry_(
...
   if (ret == SQLITE_CONSTRAINT) {
-    sqlite::Statement s2(
-        db_, "SELECT sequence, hash FROM leaves WHERE sequence = ?");
-    s2.BindUInt64(0, logged.sequence_number());
-    if (s2.Step() == SQLITE_ROW) {
-      std::string existing_hash;
-      s2.GetBlob(1, &existing_hash);
-
-      if (hash == existing_hash) {
-        return this->OK;
```
Comment (pphaneuf) [inline cpp/log/sqlite_db-inl.h]: We need to keep this case, it's just the second `SELECT` at lines 181-184 that need to go.

File: cpp/log/file_db.h [hunk 2]
```diff
@@ -87,14 +86,14 @@ class FileDB : public Database<Logged> {
...
-  // index lookups. This is also the contiguous part of the log that
-  // we are able to serve.
-  std::vector<std::string> dense_entries_;
+
+  int64_t contiguous_size_;
+  std::unordered_map<std::string, int64_t> id_by_hash_;
+
   // This is a mapping of the non-contiguous entries of the log (which
   // can happen while it is being fetched). When entries here become
   // contiguous with those in dense_entries_, they are moved there.
```
Comment (pphaneuf) [inline cpp/log/file_db.h]: This comment will need some fixing up (there's no `dense_entries_` anymore).

File: cpp/log/sqlite_db-inl.h (hunk 1 above)
Comment (AlCutter) [inline cpp/log/sqlite_db-inl.h]: I'd removed it on the broken assumption that the reason this was added (fetcher dupes) had gone away, but I've added this back in.

File: cpp/log/sqlite_db-inl.h [hunk 3]
```diff
@@ -164,6 +168,8 @@ typename Database<Logged>::WriteResult SQLiteDB<Logged>::CreateSequencedEntry_(
 
   int ret = statement.Step();
   if (ret == SQLITE_CONSTRAINT) {
+    // Check whether we're trying to store a hash/sequence pair which already
+    // exists - if it's identical we'll return OK as it could be the fetcher.
     sqlite::Statement s2(
         db_, "SELECT sequence, hash FROM leaves WHERE sequence = ?");
```
Comment (pphaneuf) [inline cpp/log/sqlite_db-inl.h]: Not quite sure why this isn't just `SELECT hash FROM ...`? Whatever.

File: cpp/log/file_db-inl.h [hunk 4]
```diff
@@ -62,36 +73,27 @@ typename Database<Logged>::WriteResult FileDB<Logged>::CreateSequencedEntry_(
...
+  std::unique_lock<std::mutex> lock(lock_);
+
   // Try to create.
-  util::Status status(cert_storage_->CreateEntry(new_hash, data));
+  util::Status status(cert_storage_->CreateEntry(seq_str, data));
   if (status.CanonicalCode() == util::error::ALREADY_EXISTS) {
-    return this->ENTRY_ALREADY_LOGGED;
+    std::string existing_data;
+    status = cert_storage_->LookupEntry(seq_str, &existing_data);
+    CHECK_EQ(status, util::Status::OK);
```
Comment (pphaneuf) [inline cpp/log/file_db-inl.h]: Did I put in a `CHECK_OK`? This is fine too (the important is that it prints the status if it's not OK).

File: Makefile.am [hunk 5]
```diff
@@ -434,7 +434,7 @@ cpp_monitor_database_test_LDADD = \
 	cpp/libcore.a \
 	cpp/libtest.a \
 	$(libevent_LIBS) \
-	-lglog -lgflags -lprotobuf -lcrypto -lsqlite3
+	-lglog -lgflags -lprotobuf -lpthread -lcrypto -lsqlite3
```
Comment (pphaneuf) [inline Makefile.am]: Should rebase. :smiley:

File: cpp/log/sqlite_db-inl.h [hunk 6]
```diff
@@ -174,14 +180,8 @@ typename Database<Logged>::WriteResult SQLiteDB<Logged>::CreateSequencedEntry_(
...
         return this->OK;
       }
-
-      return this->SEQUENCE_NUMBER_ALREADY_IN_USE;
     }
-
-    sqlite::Statement s3(db_, "SELECT hash FROM leaves WHERE hash = ?");
-    s3.BindBlob(0, hash);
-    CHECK_EQ(SQLITE_ROW, s3.Step());
-    return this->ENTRY_ALREADY_LOGGED;
```
Comment (pphaneuf) [inline cpp/log/sqlite_db-inl.h]: This enum entry isn't used anymore, should remove it from the enum.

File: cpp/log/file_db-inl.h (hunk 4 above)
Comment (AlCutter) [inline cpp/log/file_db-inl.h]: Nope :/

File: Makefile.am (hunk 5 above)
Comment (AlCutter) [inline Makefile.am]: Coming right up!

File: cpp/log/sqlite_db-inl.h (hunk 6 above)
Comment (AlCutter) [inline cpp/log/sqlite_db-inl.h]: Good call, dinged.

File: cpp/log/file_db-inl.h [hunk 7]
```diff
@@ -267,26 +272,27 @@ void FileDB<Logged>::BuildIndex() {
   // this should not be necessarily, but just to be sure...
   std::lock_guard<std::mutex> lock(lock_);
 
-  const std::set<std::string> hashes(cert_storage_->Scan());
-  dense_entries_.reserve(hashes.size());
+  const std::set<std::string> sequence_numbers(cert_storage_->Scan());
+  id_by_hash_.reserve(sequence_numbers.size());
```
Comment (pphaneuf) [inline cpp/log/file_db-inl.h]: Not sure how helpful this is, but let's keep it...

**Comment by pphaneuf** [comment]
LGTM

**Pull request merged**